README
    gKamus - Kamus Bahasa Inggris-Indonesia

    Version 1.0

    Copyright(c) 2008-2011 Ardhan Madras <ajhwb@knac.com>

    http://gkamus.sourceforge.net

================================================================================
TENTANG GKAMUS

        gKamus adalah program sederhana Kamus Bahasa Inggris - Indonesia, 
    dibuat karena rasa frutrasi saya dengan membuka-buka kamus edisi cetakan 
    setiap kali membaca artikel, ebook, atau dokumen lainnya. Program kamus ada 
    berapa yang tersedia seperti Gidic di Linux, atau program komersial Linguist 
    di Windows, tetapi kadang kala kata yang dicari tidak terdapat dalam 
    kamusnya, perlu penambahan kata, arti, dan lain-lain. Untuk memodifikasi 
    file kamus, hanya penulisnya atau orang yang bisa membaca kode program yang 
    dapat melakukan modifikasi dan menulis semacam editor, kadang mereka 
    menggunakan format biner versi mereka sendiri untuk kamusnya yang tidak 
    dapat dimodifikasi begitu saja, Anda memerlukan paling tidak semacam 
    Hex Editor.

        Lalu bagaimana dengan pengguna biasa?, yang tidak mengerti dengan bahasa 
    pemrograman, bilangan hex atau biner? mereka tentu "tidak bisa" melakukan 
    perubahan. Tidak ada gunanya membuat program yang bebas jika hanya 
    Anda (penulis) yang dapat merubahnya. Kemudian saya menulis versi program 
    kamus saya sendiri, mulanya sangat sederhana, untuk dijalankan di shell dan 
    mencari kata yang cocok di file kamusnya lalu menampilkan artinya, file 
    kamusnya berformat teks ASCII, editor untuk modikasi file kamus saya gunakan 
    vim atau ed. Masalah yang dihadapi adalah bagaimana menulis setiap kata 
    bahasa inggris dan definisinya, tidak mungkin untuk menulis semua kata dalam 
    waktu yang singkat, untungnya Firmansyah pada halaman blognya ia telah 
    mempunyai database kamus bahasa inggris yang menurut saya sudah termasuk 
    lengkap (sekitar 23000 kata dan definisi). Lalu saya menanyakan bagaimana 
    lisensi file miliknya itu, apakah tidak ada batasan untuk modifikasi dan 
    distribusi ulang? Firmansyah menjawab dan memperbolehkan saya untuk 
    menggunakan file-nya di aplikasi saya. Kemudian Firmansyah mengirimkan saya 
    aplikasi GUI untuk kamusnya yang ia tulis dengan PHP-GTK, saya pun mulai 
    menggunakan aplikasi itu di sistem saya. Namun kecepatan aplikasi itu masih 
    lambat, walaupun dengan perubahan sana-sini aplikasi itu masih berjalan 
    lambat. Saya akhirnya memutuskan untuk menulis GUI yang baru, yang cepat, 
    dan bisa berjalan dimana saja (tanpa interpreter), lalu membuat editor untuk 
    modifikasi kamus.

    Untuk membuat gKamus bisa digunakan di sistem operasi mana saja, gKamus 
    harus ditulis dengan bahasa yang bisa berjalan di mana saja, dalam hal ini 
    C dan GTK+. Pertama, C adalah bahasa semua sistem operasi, cepat dan sangat 
    portable. Kedua, GTK+ saat ini sudah di implementasikan di banyak sistem 
    operasi. Ketiga, mereka bebas, bukan gratis, tidak perduli dengan gratis 
    atau tidak, yang penting library mereka tidak ada pembatasan sana-sini untuk 
    distribusi ulang. Oke, cukup ngomongnya! 

================================================================================
FORMAT KAMUS

        gKamus menggunakan format text UNIX <LF> ASCII sebagai format kamus-nya.
    Format file kamus gKamus adalah sebagai berikut:

        <kata> <tab> <definisi> [baris baru]

        Contoh:

        kernel <tab> biji, inti, inti dari sistem operasi komputer [baris baru]

        Atau dalam format string C:

        kernel\tbiji, inti, inti dari sistem operasi komputer\n

        Kata dan definisi dipisahkan dengan sebuah dan setiap baris kata 
    dipisahkan dengan sebuah baris baru, kamus kemudian di sortir menurut abjad.
    Misalkan:

        a
        aback
        abaft
        abalone
        ...

    Penggunaan format text pada gKamus adalah sebagai berikut:

    * Format text itu simple, sehingga siapa saja bisa menggunakannya, tidak
      seperti format biner yang tertutup dan tidak bisa langsung saja digunakan
      (jika kita ingin merubahnya ke format lain)

    * Ukuran file, teks mempunyai ukuran yang jauh lebih kecil jika
      dibandingkan dengan biner.

    * Modifikasi mudah, dengan teks editor biasa.

    * Tidak seperti UNIX, Windows membedakan antara teks dan biner. Kode akan
      lebih sederhana jika menggunakan teks.

    Kekurangan format teks:

    * Kecepatan baca teks tidak secepat biner, karena alokasi jumlah bytes data
      biner tidak bisa diterapkan pada data teks sehingga pembacaan acak pada
      data teks tidak dimungkinkan. Tapi dengan teknik yang baik, kita dapat
      mengatasi hal ini.

================================================================================
INSTALASI

        gKamus dibuat dengan menimbang portabilitas, artinya gKamus dapat 
    dipakai dimana saja dengan mudah di sistem Linux (distro apa saja) atau 
    Windows. gKamus akan selalu disertakan dengan standar instalasi dan 
    kompilasi GNU (GNU Build Tools). Anda tentu tidak bisa melakukan instalasi 
    di direktori standar dengan akses non-root jika berada di sistem lain, atau 
    sistem tersebut tidak punya kompiler dan development library untuk 
    kompilasi. Untuk itu gKamus dibuat versi portable-nya dalam bentuk biner 
    untuk langsung digunakan dari harddisk atau memory flash.

    * Kompilasi di Linux

    Untuk kompilasi gKamus di Linux diperlukan development library sebagai 
    berikut:

    - GTK+ >= 2.12.8

    Lakukan pengecekan library ini dengan pkg-config:

    #pkg-config --modversion gtk+-2.0

    Jika pkg-config tidak menemukan library ini, install terlebih dahulu dengan
    installer distro Linux anda atau menggunakan online installer dari server
    repository distro anda, contoh kompilasi gKamus:

    #tar -xzvf gkamus-x.y.tar.gz #x.y adalah versi dari gkamus
    #cd gkamus-x.y
    #./configure
    #make
    #make install

    Anda juga dapat menentukan lokasi instalasi dengan:

    #./configure --prefix=/home/user/gkamus
    
    Ini akan membuat gKamus di install pada direktori $HOME user.

    Agar file kamus bisa ditulis, set permisi file-file kamus agar bisa ditulis 
    oleh user lain seperti berikut:

    #cd /usr/local/share/gkamus #direktori standar gkamus, rubah sesuai direktori anda
    #chmod 666 gkamus-en.dict gkamus-id.dict

    Untuk informasi lebih lanjut, lihat file INSTALL. Setelah kompilasi dan 
    instalasi berhasil anda dapat menjalankan gKamus dari menu atau dari shell.

    * Kompilasi di Windows

    Untuk kompilasi di Windows, anda memerlukan:

    - GCC versi Windows:
        - Mingw <http://www.mingw.org> atau,
        - Cygwin <http://www.cygwin.org>
    - GTK+ development versi Windows <http://www.gtk.org>

    Anda bisa mengambil file Makefile.win untuk kompilasi dengan Dev-C++ IDE di
    source code gKamus untuk Windows.

    Catatan: gKamus belum pernah dikompilasi di compiler lain seperti Microsoft
    Visual C++ (MSVC) dan lain-lain. Penulis tidak memiliki keinginan untuk 
    melakukan kompilasi di compiler-compiler tersebut, tapi anda bisa 
    mencobanya dengan melihat manual GTK+ tentang cross compiler.

================================================================================
LISENSI

    * Modifikasi, salin ulang, dan distribusi ulang gKamus dan gKamus editor
      diperbolehkan sesuai dengan GNU Public License (GPL) versi 2, lihat
      file COPYING

    * File kamus gKamus memakai database kamus yang aslinya dibuat oleh
      Firmansyah <leonard_gimsong@yahoo.com>, yang kemudian dijadikan format
      teks. Dia telah menyetujui database kamusnya untuk digunakan, bisa di
      distribusi ulang, dan di modifikasi ulang (keep rock'n you guy!)

================================================================================
PERUBAHAN

    Lihat file ChangeLog

================================================================================
TIP

        Meskipun file kamus gKamus bisa dimodifikasi dengan teks editor biasa
    (UNIX <LF> format) selalu gunakan gKamus untuk mengedit kamus, ini untuk
    menjaga integritas data kamus

================================================================================
PENULIS

        Lihat file AUTHORS


ChangeLog
Perubahan di 0.3:

- Tidak menggunakan fungsi sortir untuk kolum kata (versi 0.1 dan 0.2), ini yang 
  mengakibatkan pemasukan kata pada kolum menjadi sangat lambat pada GTK+ versi
  2.10.14 ke atas karena sortir row yang sangat banyak
- Tidak menggunakan libglade, glade telah direncanakan hanya untuk membuat
  interface yang dapat dipanggil dengan GtkBuilder, libglade akan ditinggalkan
- Pencarian otomatis kata yang dimasukkan pada entri (paling mendekati)
- Memakai GTK+-2.12.8
- Memakai libsexy untuk widget entry pencarian
- Menambah Irregular Verbs (kata kerja tidak beraturan)
- Perbaikan kode dan interfaces

Perubahan di 0.2:

- Menambahkan kamus Indonesia - Inggris (gkamus-id.dict)
- Mulai versi 0.2, struktur data kamus menggunakan double linked list
- Integrasi editor
- Efisiensi pencarian dengan pencarian linear di double linked list
- Efisiensi beberapa widget
- Instalasi dan kompilasi standar GNU
- Beberapa maintenance kode program

Changes in 1.0
- GTK+ 2.18.9
- Disable dictionary editing
- English localization
- Leaving libsexy since GtkEntry's icon is builtin
- Improved search algorithm


src/gui.h
/*
 *  gui.h, gKamus (http://gkamus.sourceforge.net)
 */

#ifndef __GUI_H__
#define __GUI_H__

#include <gtk/gtk.h>

#include "main.h"

enum
{
  COLUMN_WORD,
  COLUMN_N
};

extern GtkWidget *window_main;
extern GtkWidget *menu_save;
extern GtkWidget *menu_autosearch;
extern GtkWidget *menu_tool_ei;
extern GtkWidget *menu_tool_ie;
extern GtkWidget *entry_search;
extern GtkWidget *treev_word;
extern GtkWidget *textv_definition;
extern GtkWidget *window_modify;
extern GtkWidget *entry_mod;
extern GtkWidget *textv_mod;
extern GtkWidget *window_alpha;
extern GtkWidget *window_verb;
extern GtkWidget *treev_verb;
extern GtkWidget *window_tenses;
extern GtkWidget *textv_tenses;

void set_sensitive (GtkWidget*, gboolean);
gint create_dialog (GtkMessageType, gpointer, const gchar*, const gchar*, ...);
GtkWidget* create_window_main (void);
GtkWidget* create_window_modify (GkamusModifyMode);
GtkWidget* create_window_alpha (void);
GtkWidget* create_window_verb (void);
GtkWidget* create_window_tenses (void);
GtkWidget* create_dialog_about (void);
GtkWidget* create_dialog_file (gpointer, GtkFileChooserAction);

#endif /* __GUI_H__ */



src/main.h
/*
 *  main.h, gKamus (http://gkamus.sourceforge.net)
 */

#ifndef __MAIN_H__
#define __MAIN_H__

#include <gtk/gtk.h>

typedef struct
{
  gchar *word;
  gchar *definition;
} Dictionary;

typedef enum
{
  GKAMUS_MODIFY_ADD = 0,
  GKAMUS_MODIFY_EDIT,
} GkamusModifyMode;

typedef enum
{
  GKAMUS_DIC_EN = 0,
  GKAMUS_DIC_ID
} GkamusDicMode;

typedef enum
{
  GKAMUS_DIC_STATUS_ERROR = 0,
  GKAMUS_DIC_STATUS_NOT_FOUND,
  GKAMUS_DIC_STATUS_INVALID,
  GKAMUS_DIC_STATUS_MAX_ENTRY,
  GKAMUS_DIC_STATUS_OK
} GkamusDicStatus;

#define PROGRAM_NAME    "gKamus"
#define	PROGRAM_VERS    "1.0"
#define PROGRAM_DESC    "Simple English - Indonesia dictionary"
#define PROGRAM_SITE    "http://gkamus.sourceforge.net"
#define PROGRAM_ICON    "gkamus.png"
#define MAX_WORD        16
#define MAX_DEFINITION  256
#define MAX_ENTRY       100000

extern GList *dict;
extern gchar *dict_file;
extern const gchar const * dict_en;
extern const gchar const * dict_id;
extern gchar *prg_name;
extern const gchar * const dic_errmsg[];

#endif /* __MAIN_H__ */



src/callback.h
/*
 *  callback.h, gKamus (http://gkamus.sourceforge.net)
 */

#ifndef __CALLBACK_H__
#define __CALLBACK_H__

#include <gtk/gtk.h>

gboolean on_window_main_delete_event (GtkWidget*, GdkEvent*, gpointer);
void on_entry_search_icon_pressed (GtkWidget*, GtkEntryIconPosition, GdkEvent*, gpointer);
void on_entry_search_changed (GtkWidget*, gpointer);
void on_window_modify_destroy (GtkWidget*, gpointer);
void on_button_mod_ok_clicked (GtkWidget*, gpointer);
gboolean on_textv_mod_key_press_event (GtkWidget*, GdkEventKey*, gpointer);
void on_textv_mod_buffer_changed (GtkTextBuffer*, gpointer);
void select_treev_word (gint);
void on_menu_save_activate (GtkWidget*, gpointer);
void on_menu_saveas_activate (GtkWidget*, gpointer);
void on_menu_quit_activate (GtkWidget*, gpointer);
void on_menu_copy_activate (GtkWidget*, gpointer);
void on_menu_paste_activate (GtkWidget*, gpointer);
void on_menu_add_activate (GtkWidget*, gpointer);
void on_menu_edit_activate (GtkWidget*, gpointer);
void on_menu_delete_activate (GtkWidget*, gpointer);
void on_menu_tool_ei_activate (GtkWidget*, gpointer);
void on_menu_tool_ie_activate (GtkWidget*, gpointer);
void on_menu_alpha_activate (GtkWidget*, gpointer);
void on_menu_verb_activate (GtkWidget*, gpointer);
void on_menu_tenses_activate (GtkWidget*, gpointer);
void on_menu_about_activate (GtkWidget *, gpointer);
void on_treev_word_columns_changed (GtkWidget *, gpointer);
void on_button_paste_clicked (GtkWidget *, gpointer);
void on_button_find_clicked (GtkWidget *, gpointer);
gboolean on_window_alpha_delete_event (GtkWidget*, GdkEvent*, gpointer);
void on_window_verb_destroy (GtkWidget*, gpointer);
void on_window_tenses_destroy (GtkWidget*, gpointer);
void on_combob_tenses_changed (GtkComboBox*, gpointer);

#endif /* __CALLBACK_H__ */



src/function.c
/*
 *  function.c, gKamus (http://gkamus.sourceforge.net)
 *  Copyright (C) 2008-2011, Ardhan Madras <ajhwb@knac.com>
 *
 *  gKamus is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  gKamus is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with gKamus; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <gtk/gtk.h>

#include <stdlib.h>
#include <string.h>

#include "main.h"
#include "gui.h"

static gchar** divide_line (const gchar*, const gchar*);
static gboolean check_dict (gpointer);

/* 
 * Simpan file kamus.
 */
gsize
write_dict (const gchar *file)
{
  g_return_val_if_fail (file != NULL, 0);

  GIOChannel *write;

  if (!(write = g_io_channel_new_file (file, "w", NULL)))
    return 0;

  Dictionary *dic;
  GList *list = dict;
  gsize ret, bytes;
  gchar *tmp;

  static const gchar head[] =
  (
    "[gKamus File Format v1.0]\n[Encoding ANSI]\n[File ini adalah \"public"
    " domain\", tidak ada hak cipta]\n[Ditulis oleh: Firmansyah, Ardhan"
    " Madras]\n[Website http://gkamus.sourceforge.net]\n\n"
  );

  g_io_channel_write_chars (write, head, -1, &bytes, NULL);
  ret = bytes;
  while (list)
    {
      dic = list->data;
      tmp = g_strconcat (dic->word, "\t", dic->definition, "\n", NULL);
      g_io_channel_write_chars (write, tmp, -1, &bytes, NULL);
      ret += bytes;
      g_free (tmp);
      list = list->next;
    }
  g_io_channel_shutdown (write, TRUE, NULL);
  g_io_channel_unref (write);
  
  return ret;
}

/* 
 * Cek validitas format, simpel saja ;p 
 */
static gboolean
check_dict (gpointer read)
{
  g_return_val_if_fail (read != NULL, FALSE);

  GIOStatus status;
  gchar *line;
  gsize len;
  gboolean ret;

  status = g_io_channel_read_line ((GIOChannel *) read, &line, &len, NULL, NULL);
  if (status == G_IO_STATUS_ERROR)
    {
      if (len)
        g_free (line);
      return FALSE;
    }
  ret = !strcmp (line, "[gKamus File Format v1.0]\n");
  g_free (line);

  return ret;
}

/*
 * Bagi baris (line) kamus menjadi kata dan definisinya.
 * free dengan g_strfreev().
 */ 
static gchar**
divide_line (const gchar *line, 
             const gchar *delimeter)
{
  g_return_val_if_fail (line != NULL, NULL);
  g_return_val_if_fail (delimeter != NULL, NULL);

  /* cek '[' dan line kosong */
  if (*line == '[' || *line == '\n')
    return NULL;
  gchar **split = g_strsplit_set (line, delimeter, -1);
  if (g_strv_length (split) != 2)
    {
      g_strfreev (split);
      split = NULL;
    }

  return split;
}

/*
 * Cari kata di list GList (linear search).
 */
static gint
search_dict1 (const gchar *word)
{
  g_return_val_if_fail (word != NULL, -1);

  GList *list = dict;
  Dictionary *dic;
  register gint i;

  for (i = 0; list; list = list->next, i++)
    {
      dic = list->data;
      if (!strcmp (word, dic->word))
        return i;
    }

  return -1;
}

/*
 * Uncompleted linked list binary search, has better average 
 * search time than linear search.
 */
static gint
search_dict2 (const gchar *word)
{
  g_return_val_if_fail (word != NULL, -1);

  Dictionary *dic;
  GList *start, *end;
  static GList *middle_dict = NULL;
  gint cmp;
  static guint dict_len = 0;
  register gint id = 0;

  if (dict_len == 0)
    {
      dict_len = g_list_length (dict);
      middle_dict = g_list_nth (dict, dict_len >> 1);
    }

  dic = middle_dict->data;
  cmp = strcmp (word, dic->word);
  if (cmp == 0)
    return dict_len >> 1;

  else if (cmp < 0)
    {
      start = dict;
      end = middle_dict;
    }
  else
    {
      start = middle_dict->next;
      end = NULL;
      id = (dict_len >> 1) + 1;
    }

  while (start != end)
    {
      dic = start->data;
      if (!strcmp (word, dic->word))
        return id;
      start = start->next;
      id++;
    }

  return -1;
}

gint
search_dict (const gchar *word)
{
  return search_dict2 (word);
}

/*
 * Fungsi perbandingan list untuk sortir.
 */
gint
list_compare (const Dictionary *a, 
              const Dictionary *b)
{
  return strcmp (a->word, b->word);
}

void
free_elements (Dictionary *dic, 
               gpointer    data)
{
  g_return_if_fail (dic != NULL);
  g_free (dic->word);
  g_free (dic->definition);
  g_free (dic);
}

/*
 * Load kamus ke memori (GList).
 */
GkamusDicStatus
load_dict (const gchar *file)
{
  g_return_val_if_fail (file != NULL, GKAMUS_DIC_STATUS_NOT_FOUND);

  GIOChannel *read;
  gchar *temp;
  gboolean test;

#ifdef PACKAGE_DATA_DIR
  temp = g_build_filename (PACKAGE_DATA_DIR, file, NULL);
#else
  gchar *current_dir = get_get_current_dir ();
  temp = g_build_filename (current_dir, "share", "data", file, NULL);
  g_free (current_dir);
#endif

  test = g_file_test (temp, G_FILE_TEST_IS_REGULAR | G_FILE_TEST_EXISTS);
  if (!test)
    {
      g_free (temp);
      return GKAMUS_DIC_STATUS_NOT_FOUND;
    }
  read = g_io_channel_new_file (temp, "r", NULL);
  if (!read)
    {
      g_free (temp);
      return GKAMUS_DIC_STATUS_ERROR;
    }
  if (!check_dict (read)) /* format valid ? */
    {
      g_free (temp);
      g_io_channel_shutdown (read, TRUE, NULL);
      g_io_channel_unref (read);
      return GKAMUS_DIC_STATUS_INVALID;
    }

  gchar **split;
  gchar *line, *word, *definition;
  guint i = 0;

  /* free list dan element */
  if (dict != NULL)
    {
      g_list_foreach (dict, (GFunc) free_elements, NULL);
      g_list_free (dict);
      dict = NULL;
    }
  while (g_io_channel_read_line (read, &line, NULL, NULL, NULL) != G_IO_STATUS_EOF)
    {
      if (i > MAX_ENTRY) /* periksa entri maksimal */
        {
          g_io_channel_shutdown (read, TRUE, NULL);
          g_io_channel_unref (read);
          g_free (line);
          g_free (temp);
          return GKAMUS_DIC_STATUS_MAX_ENTRY;
        }
      if ((split = divide_line (line, "\t")))
        {
          Dictionary *dic = g_new (Dictionary, 1);
          word = g_ascii_strdown (g_strstrip (split[0]), -1);
          definition = g_strstrip (split[1]);
          dic->word = g_strdup (word);
          dic->definition = g_strdup (definition);
          g_free (word);

          /* g_list_insert() atau g_list_append() tentu akan lama untuk
           * data yang banyak, disini kita masukkan @dic pada head (awal)
           * list sort list sesuai abjad
           */
          dict = g_list_prepend (dict, dic);
          i++;
       }
      g_free (line);
      g_strfreev (split);
    }
  /* reverse list dengan sort, kita tidak perlu g_list_reverse() */
  dict = g_list_sort (dict, (GCompareFunc) list_compare);

  /* set @dict_file */
  if (dict_file)
    g_free (dict_file);
  dict_file = g_strdup (temp);
  g_free (temp);

  g_io_channel_shutdown (read, TRUE, NULL);
  g_io_channel_unref (read);

  return GKAMUS_DIC_STATUS_OK;
}

/*
 * Masukkan kata ke treeview.
 */
void
set_treev_word (void)
{
  GList *ptr;
  Dictionary *d;
  GtkTreeModel *model;
  GtkTreeIter iter;

  model = gtk_tree_view_get_model ((GtkTreeView *) treev_word);
  gtk_list_store_clear ((GtkListStore *) model);
  ptr = dict;
  while (ptr)
    {
      d = ptr->data;
      gtk_list_store_append ((GtkListStore *) model, &iter);
      gtk_list_store_set ((GtkListStore *) model, &iter, COLUMN_WORD, d->word, -1);
      ptr = g_list_next (ptr);
    }
}

void
show_dicstatus (GkamusDicStatus status)
{
  static const gchar msg[] = "Silahkan perbaiki kesalahan pada kamus atau "
                             "lakukan instalasi ulang bila perlu.";
  create_dialog (GTK_MESSAGE_ERROR, window_main, dic_errmsg[status], msg);
}



src/main.c
/*
 *  main.c, gKamus (http://gkamus.sourceforge.net)
 *  Copyright (C) 2008-2011, Ardhan Madras <ajhwb@knac.com>
 *
 *  gKamus is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  gKamus is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with gKamus; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/*
 *  18-Feb-2008, 13.32 WIB, Dream Theater "Perfect Stranger".
 */

#include <gtk/gtk.h>

#ifdef G_OS_UNIX
# include <sys/types.h>
# include <sys/stat.h>
# include <sys/file.h>
# include <fcntl.h>
# include <unistd.h>
# include <errno.h>
#endif

#include "main.h"
#include "gui.h"
#include "callback.h"
#include "function.h"

GList *dict = NULL; /* struktur data kamus (double linked list) */
gchar *dict_file = NULL; /* path absolute kamus yang sedang aktif */
const gchar const * dict_en	= "gkamus-en.dict";	/* nama file kamus inggris */
const gchar const * dict_id	= "gkamus-id.dict";	/* nama file kamus indonesia */
gchar *prg_name = NULL; /* string untuk argv[1] */

const gchar * const dic_errmsg[] =
{
  "File kamus tidak dapat dibuka",
  "File kamus tidak ditemukan",
  "Format file salah atau tidak diketahui",
  "Jumlah entri kamus melewati batas maksimal",
  "File kamus ok"
};


gint
main (gint   argc, 
      gchar *argv[])
{
#ifdef G_OS_UNIX
  gint fd, ret;
  
  fd = open ("/tmp/gkamus.lock", O_CREAT | O_RDWR, 0666);
  if (fd == -1)
    {
      g_critical ("open: %s.\n", g_strerror (errno));
      return -1;
    }
  ret = flock (fd, LOCK_NB | LOCK_EX);
  if (ret == -1)
    {
      if (errno != EWOULDBLOCK)
        {
          g_critical ("flock: %s.\n", g_strerror (errno));
        }
      close (fd);
      return 0;
    }
#endif
  gtk_init (&argc, &argv);

  /* set prg_name ke argv[0] */
  prg_name = argv[0];

  window_main = create_window_main (); /* buat window utama gkamus */
  window_alpha = create_window_alpha ();

  /* load file kamus */
  gint status = load_dict (dict_en);
  if (status != GKAMUS_DIC_STATUS_OK)
    {
      show_dicstatus (status);
      return -1;
    }
  set_treev_word ();
  select_treev_word (0);
  if (gtk_events_pending ())
    gtk_main_iteration ();
  gtk_widget_show_all (window_main);
  gtk_main ();

  /* gtk_widget_destroy (window_main); */
  if (window_alpha)
    gtk_widget_destroy (window_alpha);
  if (window_verb)
    gtk_widget_destroy (window_verb);
  g_list_foreach (dict, (GFunc) free_elements, NULL);
  g_list_free (dict);
#ifdef G_OS_UNIX
  ret = flock (fd, LOCK_UN);
  if (ret == -1)
    g_critical ("flock: %s.\n", g_strerror (errno));    
  close (fd);
#endif

  return 0;
}
